File: paack-java-sdk/src/main/java/eu/paack/sdk/api/RetailerLocationApi.java
package eu.paack.sdk.api;

import com.fasterxml.jackson.core.type.TypeReference;
import eu.paack.sdk.PaackConstants;
import eu.paack.sdk.api.model.request.RetailerGetRequest;
import eu.paack.sdk.api.model.response.Error;
import eu.paack.sdk.api.model.response.PaackResponse;
import eu.paack.sdk.api.model.response.RetailerGetResponse;
import eu.paack.sdk.api.model.response.RetailerLocationResponse;
import eu.paack.sdk.api.validator.RetailerLocationValidator;
import eu.paack.sdk.exceptions.ApiException;
import eu.paack.sdk.model.RetailerLocation;
import lombok.Builder;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.NameValuePair;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

import static eu.paack.sdk.api.PaackEndpoint.retailer;

/**
 * Retailer Location Api is a public service available for retailers to manage their locations (store / warehouse addresses).
 * You can use it to create, retrieve, update and delete locations.
 *
 * More details can be found in the Paack API documentation: https://paack.readme.io/reference/orders-tracking-requests
 */
@Slf4j
@SuperBuilder
@NoArgsConstructor
public class RetailerLocationApi extends PaackApi {

    @Builder.Default
    private RetailerLocationValidator retailerRequestValidator = new RetailerLocationValidator();

    /**
     * Creates a new retailer location with a name, alias, type (store / warehouse) and address.
     * Args:
     * payload: The retailer location data to create
     * Returns:
     * RetailerLocationResponse | Error
     * @param request
     * @return
     */
    public PaackResponse<RetailerLocationResponse, Error> create(RetailerLocation request) {

        Optional<Error> error = retailerRequestValidator.checkForErrors(request);
        if(error.isPresent()) {
            return errorMessage(error.get());
        }

        try {
            PaackResponse<RetailerLocationResponse, Error> response = apiClient.invokeAPI(retailer,
                    "POST",
                    null,
                    null,
                    request,
                    new TypeReference<RetailerLocationResponse>() {
                    });
            log.info(response.toString());
            return PaackResponse.<RetailerLocationResponse, Error>builder()
                    .data(response.getData())
                    .error(response.getError())
                    .build();
        } catch (ApiException e) {
            log.error("Failed to create the retailer location.", e);
            return errorMessage("RetailerLocationApi.create", e.getMessage());
        }
    }

    /**
     * Remove a retailer location.
     *
     * Args:
     * retailerLocationId: The retailer location ID to update
     * Returns:
     * DeleteOrderResponse | ErrorResponse
     * @param retailerLocationId
     * @return
     */
    public PaackResponse<RetailerLocation, Error> delete(String retailerLocationId) {
        if (retailerLocationId == null || retailerLocationId.length() == 0) {
            return errorMessage("RetailerId cannot be empty", "retailerId", "001");
        }
        try {
            PaackResponse<RetailerLocation, Error> response = apiClient.invokeAPI(retailer,
                    "DELETE",
                    Collections.singletonList(param(PaackConstants.PARAM_RETAILER_LOCATION_ID, retailerLocationId)),
                    null,
                    null,
                    new TypeReference<RetailerLocation>() {
                    });
            log.info(response.toString());
            return PaackResponse.<RetailerLocation, Error>builder()
                    .data(response.getData())
                    .build();
        } catch (ApiException e) {
            log.error("Failed to delete the retailer location.", e);
            return errorMessage("RetailerLocationApi.delete", e.getMessage());
        }
    }

    /**
     * Update an existing retailer location.
     *
     * Args:
     * retailerLocationId: The retailer location ID to update
     * payload: The retailer location data to update
     * Returns:
     * RetailerLocationResponse | Error
     * @param request
     * @return
     */
    public PaackResponse<RetailerLocationResponse, Error> patch(RetailerLocation request) {
        if (request.getId()== null || request.getId().length() == 0) {
            return errorMessage("Retailer location id cannot be empty", "Id", "001");
        }
        try {
            PaackResponse<RetailerLocationResponse, Error> response = apiClient.invokeAPI(retailer,
                    "PATCH",
                    Collections.singletonList(param(PaackConstants.PARAM_RETAILER_LOCATION_ID, request.getId())),
                    null,//Collections.singletonList(params),
                    request,
                    new TypeReference<RetailerLocationResponse>() {
                    });
            log.info(response.toString());
            return PaackResponse.<RetailerLocationResponse, Error>builder()
                    .data(response.getData())
                    .error(response.getError())
                    .build();
        } catch (ApiException e) {
            log.error("Failed to update the retailer location.", e);
            return errorMessage("RetailerLocationApi.patch", e.getMessage());
        }
    }

    /**
     * Retrieve a list of retailer locations.
     *
     * Args:
     * GetRetailerLocationRequest
     * Returns:
     * GetRetailerLocationResponse | ErrorResponse
     * @param request
     * @return
     */
    public PaackResponse<List<RetailerLocation>, Error> get(RetailerGetRequest request){
        try {
            PaackResponse<RetailerGetResponse, Error> response = apiClient.invokeAPI(retailer,
                    "GET",
                    null,
                    mapQueryParams(request),
                    null,
                    new TypeReference<RetailerGetResponse>() {
                    });
            log.info(response.toString());
            return PaackResponse.<List<RetailerLocation>, Error>builder()
                    .data(response.getData() == null ? null : response.getData().getLocations())
                    .error(response.getError())
                    .build();
        } catch (ApiException e) {
            log.error("Failed to retrieve the retailer locations.", e);
            return errorMessage("RetailerLocationApi.get", e.getMessage());
        }
    }

    private List<NameValuePair> mapQueryParams(RetailerGetRequest request) {
        List<NameValuePair> pairList = new ArrayList<>();
        if (request.getRetailerId() != null) {
            pairList.add(param("retailer_id", request.getRetailerId()));
        }

        if (request.getId() != null) {
            pairList.add(param("id", request.getId()));
        }

        if (request.getRetailerName() != null) {
            pairList.add(param("retailer_name", request.getRetailerName()));
        }

        if (request.getLocationName() != null) {
            pairList.add(param("location_name", request.getLocationName()));
        }

        if (request.getCity() != null) {
            pairList.add(param("city", request.getCity()));
        }

        if (request.getPostcode() != null) {
            pairList.add(param("postcode", request.getPostcode()));
        }

        if (request.getAlias() != null) {
            pairList.add(param("alias", request.getAlias()));
        }

        if (request.getType() != null) {
            pairList.add(param("type", request.getType()));
        }
        return pairList;
    }


}

File: paack-go-sdk/rest/retailer_location_api.go
package rest

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/PaackEng/paack-go-sdk/api_client"
	"github.com/PaackEng/paack-go-sdk/logger"
	"github.com/PaackEng/paack-go-sdk/mixin"
	"github.com/PaackEng/paack-go-sdk/types"
	"github.com/PaackEng/paack-go-sdk/types/responses"
	"io"
	"strconv"
)

/*
RetailerLocationApi - is a public service available for retailers.
You can use it to manage the retailers' locations.

More details can be found in the Paack API documentation: https://paack.readme.io/reference/orders-tracking-requests
*/
type RetailerLocationApi struct {
	BaseApi
}

func NewRetailerLocationApi(audience, domain string, path *mixin.Resources, client api_client.ApiClient, logger logger.Logger) *RetailerLocationApi {

	return &RetailerLocationApi{
		*NewBaseApi(client, domain, audience, path, logger),
	}
}

/*
GetRetailerLocation - Retrieve a list of retailer locations

Args:

	GetRetailerLocationRequest

Returns:

	GetRetailerLocationResponse | ErrorResponse
*/
func (r *RetailerLocationApi) GetRetailerLocation(request GetRetailerLocationRequest) (successResponse *responses.GetRetailerLocationResponse, errorResponse *responses.ErrorResponse) {
	r.logger.Info("Calling RetailerLocationApi.GetRetailerLocation method")

	if len(request.RetailerID) == 0 {
		errorResponse = responses.NewError(fmt.Sprintf("retailerId cannot be empty"), "001", "RetailerID", r.logger)
		return
	}

	url := r.computeGetRetailerLocationUrl(request)

	resp, err := r.client.Get(url, true, r.audience)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		errorResponse = responses.NewError(fmt.Sprintf("Received error: %s", resp.Status), strconv.Itoa(resp.StatusCode), "", r.logger)
		return
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while reading response body: %s", err), "", "", r.logger)
		return
	}

	err = json.Unmarshal(body, &successResponse)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while unmarshalling response body: %s", err), "", "", r.logger)
		return
	}

	return
}

/*
CreateRetailerLocation - Creates a new retailer location with a name, alias, type (store / warehouse) and address.

Args:

	payload: The retailer location data to create

Returns:

	RetailerLocationResponse | ErrorResponse
*/
func (r *RetailerLocationApi) CreateRetailerLocation(payload types.RetailerLocation) (successResponse *responses.RetailerLocationResponse, errorResponse *responses.ErrorResponse) {
	r.logger.Info("Calling RetailerLocationApi.CreateRetailerLocation method")

	err, source, code := payload.Validate()
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while validating payload: %s", err), code, source, r.logger)
		return
	}

	payloadBytes := new(bytes.Buffer)
	err = json.NewEncoder(payloadBytes).Encode(payload)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while encoding payload: %s", err), "", "", r.logger)
		return
	}

	url := fmt.Sprintf("%s%s", r.domain, "/rls")
	resp, err := r.client.Post(url, payloadBytes.Bytes(), true, r.audience)

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while reading response body: %s", err), "", "", r.logger)
		return
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		errorResponse = responses.NewError(fmt.Sprintf("Received error: %s", body), strconv.Itoa(resp.StatusCode), "", r.logger)
		return
	}

	err = json.Unmarshal(body, &successResponse)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while unmarshalling response body: %s", err), "", "", r.logger)
		return
	}

	return
}

/*
UpdateRetailerLocation - Update an existing retailer location

Args:

	retailerLocationId: The retailer location ID to update
	payload: The retailer location data to update

Returns:

	RetailerLocationResponse | ErrorResponse
*/
func (r *RetailerLocationApi) UpdateRetailerLocation(retailerLocationId string, payload PatchRetailerLocationRequest) (successResponse *responses.RetailerLocationResponse, errorResponse *responses.ErrorResponse) {
	r.logger.Info("Calling RetailerLocationApi.UpdateRetailerLocation method")

	// validate retailerId
	if len(retailerLocationId) == 0 {
		errorResponse = responses.NewError(fmt.Sprintf("retailerLocationId cannot be empty"), "", "", r.logger)
		return
	}

	// validate request
	if payload.Address != nil {
		err, source, code := payload.Address.Validate()
		if err != nil {
			errorResponse = responses.NewError(fmt.Sprintf("Error while validating address: %s", err), code, "Address."+source, r.logger)
			return
		}
	}

	// create payload
	payloadBytes := new(bytes.Buffer)
	err := json.NewEncoder(payloadBytes).Encode(payload)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while encoding payload: %s", err), "", "", r.logger)
		return
	}

	// send the update request
	url := fmt.Sprintf("%s%s/%s", r.domain, "/rls", retailerLocationId)
	resp, err := r.client.Patch(url, payloadBytes.Bytes(), true, r.audience)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		errorResponse = responses.NewError(fmt.Sprintf("Received error: %s", resp.Status), strconv.Itoa(resp.StatusCode), "", r.logger)
		return
	}

	// read the response
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while reading response body: %s", err), "", "", r.logger)
		return
	}

	err = json.Unmarshal(body, &successResponse)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while unmarshalling response body: %s", err), "", "", r.logger)
		return
	}

	return
}

/*
DeleteRetailerLocation - Remove a retailer location

Args:

	retailerLocationId: The retailer location ID to update

Returns:

	DeleteOrderResponse | ErrorResponse
*/
func (r *RetailerLocationApi) DeleteRetailerLocation(retailerLocationId string) (successResponse *responses.DeleteOrderResponse, errorResponse *responses.ErrorResponse) {
	r.logger.Info("Calling RetailerLocationApi.DeleteRetailerLocation method")

	// validate retailerLocationId
	if len(retailerLocationId) == 0 {
		errorResponse = responses.NewError(fmt.Sprintf("retailerLocationId cannot be empty"), "", "", r.logger)
		return
	}

	// send the delete request
	url := fmt.Sprintf("%s%s/%s", r.domain, "/rls", retailerLocationId)
	resp, err := r.client.Delete(url, true, r.audience)
	// read the response
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while reading response body: %s", err), "", "", r.logger)
		return
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		errorResponse = responses.NewError(fmt.Sprintf("Received error: %s", body), strconv.Itoa(resp.StatusCode), "", r.logger)
		return
	}

	if resp.StatusCode == 204 || resp.StatusCode >= 300 {
		return
	}

	err = json.Unmarshal(body, &successResponse)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while unmarshalling response body: %s", err), "", "", r.logger)
		return
	}

	return
}

func (r *RetailerLocationApi) computeGetRetailerLocationUrl(request GetRetailerLocationRequest) string {
	url := fmt.Sprintf("%s%s%s", r.domain, "/rls?retailer_id=", request.RetailerID)

	if len(request.ID) > 0 {
		url += fmt.Sprintf("&id=%s", request.ID)
	}

	if len(request.RetailerName) > 0 {
		url += fmt.Sprintf("&retailer_name=%s", request.RetailerName)
	}

	if len(request.LocationName) > 0 {
		url += fmt.Sprintf("&location_name=%s", request.LocationName)
	}

	if len(request.Country) > 0 {
		url += fmt.Sprintf("&country=%s", request.Country)
	}

	if len(request.City) > 0 {
		url += fmt.Sprintf("&city=%s", request.City)
	}

	if len(request.PostCode) > 0 {
		url += fmt.Sprintf("&post_code=%s", request.PostCode)
	}

	if len(request.DcID) > 0 {
		url += fmt.Sprintf("&dc_id=%s", request.DcID)
	}

	if len(request.Alias) > 0 {
		url += fmt.Sprintf("&alias=%s", request.Alias)
	}

	if len(request.Type) > 0 {
		url += fmt.Sprintf("&type=%s", request.Type)
	}

	return url
}
